crates/sui-move-build/src/linters/redundant_custom_call.rs

```diff
@@ -0,0 +1,267 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+//! This analysis flags potential custom implementations of transfer/share/freeze calls on objects
+//! that already have a store ability and where "public" variants of these calls can be used. This
+//! can be dangerous as custom transfer/share/freeze operation is becoming unenforceable in this
+//! situation.  A function is considered a potential custom implementation if it takes as a
+//! parameter an instance of a struct type defined in a given module with a store ability and passes
+//! it as an argument to a "private" transfer/share/freeze call.
+
+use move_ir_types::location::*;
+
+use move_compiler::{
+    cfgir::{
+        absint::JoinResult,
+        ast::Program,
+        visitor::{
+            LocalState, SimpleAbsInt, SimpleAbsIntConstructor, SimpleDomain, SimpleExecutionContext,
+        },
+        CFGContext, MemberName,
+    },
+    diag,
+    diagnostics::{
+        codes::{custom, DiagnosticInfo, Severity},
+        Diagnostic, Diagnostics,
+    },
+    hlir::ast::{
+        BaseType_, Command, Exp, LValue, Label, ModuleCall, SingleType, SingleType_, Type,
+        TypeName_, Type_, Var,
+    },
+    parser::ast::Ability_,
+    shared::{CompilationEnv, Identifier},
+};
+use std::collections::BTreeMap;
+
+use super::INVALID_LOC;
+
+const TRANSFER_FUN: &str = "transfer";
+const SHARE_FUN: &str = "share_object";
+const FREEZE_FUN: &str = "freeze_object";
+
+const PRIVATE_OBJ_FUNCTIONS: &[(&str, &str, &str)] = &[
```

Comment:
Not for this PR, but should we have a central place where all of these functions (and other special names) are defined? e.g., the sui-verifier also uses some of these names, and it feels like a bit of a code smell to have them sprinkled about the codebase instead of being defined in one central module.